crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+            .tables
+            .root_state_hash_by_checkpoint
+            .iter()
+            .skip_to_last()
+            .next()
+            .map(|(highest, hash)| (highest.saturating_add(1), hash))
+            .unwrap_or((0, Accumulator::default()));
+
+        for i in next_to_accumulate..=last_checkpoint {
+            let acc = self
+                .store
+                .tables
+                .state_hash_by_checkpoint
+                .get(&i)?
+                .unwrap_or_else(|| {
+                    panic!("Accumulator for checkpoint sequence number {i:?} not present in store")
+                });
+            root_state_hash.union(&acc);
+        }
+
+        // We want to enforce that this table is append only, otherwise we
+        // may end up re-accumulating the same state
+        assert!(
+            !self
+                .store
+                .tables
+                .root_state_hash_by_checkpoint
+                .contains_key(&last_checkpoint)?,
+            "StateAccumulator: root state hash already exists for checkpoint {last_checkpoint:?}"
+        );
+
+        // Update root_state_hash tables atomically
+        let batch = self.store.tables.root_state_hash_by_checkpoint.batch();
+
+        let batch = batch.insert_batch(
+            &self.store.tables.root_state_hash_by_checkpoint,
+            std::iter::once((&last_checkpoint, root_state_hash.clone())),
+        )?;
+        let batch = batch.insert_batch(
+            &self.store.tables.root_state_hash_by_epoch,
+            std::iter::once((&epoch, root_state_hash.clone())),
+        )?;
+
+        batch.write()?;
+        self.store
+            .root_state_notify_read
+            .notify(&epoch, &root_state_hash);
+
+        Ok(())
+    }
+}
+
+fn hash_checkpoint_effects(
+    state: State,
+    store: Arc<StateAccumulatorStore>,
+) -> SuiResult<Accumulator> {
+    let seq_num = state.checkpoint_seq_num;
+
+    if store
+        .tables
+        .state_hash_by_checkpoint
+        .contains_key(&seq_num)
+        .unwrap()
+    {
+        return Err(SuiError::from(
+            format!("StateAccumulator: checkpoint already hashed: {seq_num:?}").as_str(),
+        ));
+    }
+
+    let mut acc = Accumulator::default();
+
+    acc.insert_all(
+        state
+            .effects
+            .iter()
+            .flat_map(|fx| {
+                fx.created
+                    .clone()
+                    .into_iter()
+                    .map(|(obj_ref, _)| obj_ref.2)
+                    .collect::<Vec<ObjectDigest>>()
+            })
+            .collect::<Vec<ObjectDigest>>(),
+    );
+    acc.remove_all(
+        state
+            .effects
+            .iter()
+            .flat_map(|fx| {
+                fx.deleted
+                    .clone()
+                    .into_iter()
+                    .map(|obj_ref| obj_ref.2)
+                    .collect::<Vec<ObjectDigest>>()
+            })
+            .collect::<Vec<ObjectDigest>>(),
```

Comment:
possible optimization: we could build a set of the creations, then for any delete that is in creations, remove it from both (basically to avoid inserting and immediately removing a given object ref).

this may not be worth it if there are few causally-dependent txns in most checkpoints.

Reply:
Anything to get the latency of this operation down is probably worth doing. Will do

Reply:
please see my comments re the benchmark - i think we probably don't need to worry about this right now.